### ISSUE TITLE
Flatpak Support

### ISSUE THREAD

**Issue by Doomsdayrs** [ruffle-rs/ruffle#9220]
**TL;DR Status**: the current Flathub submission:
* https://github.com/flathub/flathub/pull/5473

## Describe the feature

While ruffle is still in active development and not yet stable, it is still quite useful for users to play many flash games (such as BTD) that do operate well with the current support.

To make it easier for users on Linux distributions to use Ruffle, it would be nice to get ruffle bundled with flatpak and published to flathub.

[This](https://github.com/Doomsdayrs/flathub/tree/rs.ruffle.Ruffle) repository contains my current bundling of ruffle. There are issues I want to first solve before I can submit to flathub, of course with permission from ruffle-rs to do so.

## Issues

Most of these are all XDG compliance related.

- [x] Linux packaging requirements
These are some requirements I would like to get put upstream.
  - [x] Desktop file https://github.com/ruffle-rs/ruffle/pull/11227
  - [x] Metainfo file https://github.com/ruffle-rs/ruffle/pull/11227
  - [x] SVG Logo https://github.com/ruffle-rs/ruffle/pull/11226
- [x] Immediate crash due to clipboard errors, detailed below; Dealt with by using x11 fallback.

**Comment by dginovker** [comment]
Clipboard crash is just missing permissions iirc. A good test would be to download Flatseal and enable all the permissions and see if it still crashes.

The SVG logo isn't required but highly encouraged, same thing with Desktop/metainfo file being upstreamed but it can be in your Flatpak repo. The only thing that is 100% required when submitting to Flathub is you have permission from the ruffle-rs maintainers

**Comment by Doomsdayrs** [comment]
> Clipboard crash is just missing permissions iirc. A good test would be to download Flatseal and enable all the permissions and see if it still crashes.

Apologies, I doubt that the issue is with permissions. Please do follow the mention hole with https://github.com/YaLTeR/wl-clipboard-rs/issues/32 to see. It seems it might be improper usage of libraries.

Specifically flatpak support might have to wait on https://github.com/rust-windowing/winit/pull/2156 for proper clipboard handling.

> The only thing that is 100% required when submitting to Flathub is you have permission from the ruffle-rs maintainers

That is why I am here :D

**Comment by Doomsdayrs** [comment]
One solution is to allow Clipboard to become null, and simply warn in console that it will not operate. But I despise this type of hacky solution and would ruin user experience.

**Comment by n0samu** [comment]
I understand that the suggestion is to use Smithay Clipboard, but this is not a cross-platform library like arboard, and I'm not sure the Ruffle devs would want to add this additional dependency.

Anyway @Doomsdayrs thanks for your work on this! Here's the Ruffle icon SVG:
![favicon](https://user-images.githubusercontent.com/71368227/213593797-d941644a-6491-46c2-9186-77a11c9d587d.svg)

**Comment by Doomsdayrs** [comment]
> I understand that the suggestion is to use Smithay Clipboard, but this is not a cross-platform library like arboard, and I'm not sure the Ruffle devs would want to add this additional dependency.

It can be done via abstraction. Using a separate library for Linux while retaining the current library for other platforms. In fact, the abstraction would be quite simple as the clipboard is only used in two lines of code.
 
> Anyway @Doomsdayrs thanks for your work on this! Here's the Ruffle icon SVG: ![favicon](https://user-images.githubusercontent.com/71368227/213593797-d941644a-6491-46c2-9186-77a11c9d587d.svg)

Where is this file hosted on the github?

**Comment by n0samu** [comment]
> Where is this file hosted on the github?

I don't think it is. One of the other Ruffle developers sent me this file on Discord. I do think it would be a good idea to include it in the repo somewhere though.

**Comment by WingofaGriffin** [comment]
Would love to see a Flatpak or an AppImage supported, as on readonly systems like SteamOS, we are missing libraries to run it:

```sh
./ruffle: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or directory
```

**Comment by Doomsdayrs** [comment]
> Would love to see a Flatpak or an AppImage supported, as on readonly systems like SteamOS, we are missing libraries to run it:
> 
> ```shell
> ./ruffle: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or directory
> ```

Sorry for the late reply.
Currently I am tied up with College work, hopefully once all is done I will once again contact all the various developers to work on getting support working.

**Comment by Doomsdayrs** [comment]
Alright, Now using the `nightly-2023-05-26`

Congrats to the new UI, Looks much better!

The issue with the clipboard still seems present.

![image](https://github.com/ruffle-rs/ruffle/assets/38189170/bb5cf518-8537-43ed-890c-05f3d277e06f)

---

To update, @YaLTeR from `wl-clipboard-rs` states the [following](https://github.com/YaLTeR/wl-clipboard-rs/issues/32#issuecomment-1397152379). 

To summarize, `wl-clipboard-rs` is for terminal or background applications; [arboard](https://github.com/1Password/arboard) is in the wrong for using it for GUI applications (?).

GUI apps should be using the wayland offer, which should be in `winit`, but the support has not been added in yet it [seems](https://github.com/rust-windowing/winit/pull/2156).

@YaLTeR suggests [smithay-clipboard](https://github.com/Smithay/smithay-clipboard), but that seems a bit unmaintained.

I personally want to say clipboard should be disabled on flatpak & wayland until proper support is added, but I believe that might be contrary to the goals Ruffle has currently.

**Comment by Doomsdayrs** [comment]
On a brighter note, I can run it under flatpak, but only under x11, which is sub-optimal due to many reasons I will not get too deep into.

**Comment by torokati44** [comment]
Does https://github.com/ruffle-rs/ruffle/pull/14362 help with this, as suggested by @n0samu?

**Comment by Doomsdayrs** [comment]
> Does #14362 help with this, as suggested by @n0samu?

Yep!

**Comment by zefr0x** [comment]
> The only thing that is 100% required when submitting to Flathub is you have permission from the ruffle-rs maintainers

It isn't, but it is a recommended, especially if there are problems that need to be solved in up-stream.

**Comment by Doomsdayrs** [comment]
> Does #14362 help with this, as suggested by @n0samu?

It seems that I did not fully test it, https://github.com/ruffle-rs/ruffle/issues/15213

**Comment by evilpie** [comment]
See https://github.com/flathub/flathub/pull/5473

**Comment by kjarosh** [comment]
We are now officially available on Flathub: https://flathub.org/apps/rs.ruffle.Ruffle :tada:
